lib/ui/compositing.dart

```diff
@@ -51,34 +158,74 @@ class SceneBuilder extends NativeFieldWrapperClass2 {
   SceneBuilder() { _constructor(); }
   void _constructor() native 'SceneBuilder_constructor';
 
+  // Layers used in this scene.
+  //
+  // The key is the layer used. The value is the description of what the layer
+  // is used for, e.g. "pushOpacity" or "addRetained".
+  Map<EngineLayer, String> _usedLayers = <EngineLayer, String>{};
```

Comment:
Could we also catch the case where a layer is marked an oldLayer, and thus freed, but then it's reused in another scene builder?

Reply:
That's actually a supported case in Flutter for web. The old layer becomes "released" and when you use it again the engine code will "revive" it. Other than the DOM nodes the information stores on the layers is immutable and so we can always re-inflate a DOM tree from discarded layers.

The usage you describe was not why we implemented it though. We implemented it because you can push a new layer in builder1, then not use that layer in builder2 (thus another layer could "adopt" its DOM nodes), and call `addRetained` or pass it as `oldLayer` in builder3. We even [have a test for this](https://github.com/flutter/flutter_web/blob/61dfe9fc74f30c7f19593f3d48922064bff2556f/packages/flutter_web_ui/test/engine/surface/surface_test.dart#L183).

Reply:
The test/use case you referred seems reasonable. But is it legal to do the following?
1. Create a `layer1` in frame/builder 1,
2. Call `pushLayer(..., oldLayer: layer1)` in frame/builder 2,
3. Call `addRetained(layer1)` in frame/builder 3

That sounds a little counter-intuitive and some engine implementation may view that as illegal. If we want to make this kind of usages legal, maybe it's worth to document it, and put a unit test here to ensure that all engine implementations can support it.

Reply:
Intuitively you can think of it as either of the following two ways, depending on whether the layers are immutable or mutable:

1. _Immutable_: an `EngineLayer` is simply a description of what should appear in a given subtree of a scene. It should be OK to go back and forth between two descriptions, all while updating existing underlying resources.
1. _Mutable_: `pushLayer` may actually return the same object as `oldLayer`. Since the object is the same, you'd have to pass it as `oldLayer` multiple times to communicate an "update".

Reply:
However, I don't mind putting a restriction in place. I don't think we are using layers in a way where this matters today. LMK which way you prefer.

Reply:
Let's restrict it now as I think it's easier to add capabilities than to remove capabilities that are being used :)

Reply:
SGTM. Done.